### PR TITLE
[MLIR][LLVM] Add ftz and fuse FP ops related function attribute support

### PR DESCRIPTION
**Pull request by runseny**
Adds  `denormal-fp-math-f32`, `denormal-fp-math`, `fp-contract` to llvmFuncOp attributes.

`denormal-fp-math-f32` and `denormal-fp-math` can enable the ftz, that is , flushing denormal to zero.

`fp-contract` can enable the fma fusion such as `mul + add -> fma`

### CODE REVIEW

**Comment by runseny** [comment]
@asb @zero9178 Could you help review this mr? This mr is similar with the previous work: https://github.com/llvm/llvm-project/pull/79812, Thanks in advance!

File: mlir/test/Target/LLVMIR/Import/function-attributes.ll
Comment (gysit) [inline mlir/test/Target/LLVMIR/Import/function-attributes.ll:351]: Since these are string attributes there is probably not much value in testing different string values for every attribute (i.e. ieee, preserve-sign)? Or is there specific reason for testing the different strings? Instead I would have a test for every attribute with possible different strings? So three tests here and three tests in the file below which exercises the lowering to LLVM IR.

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleImport.cpp:1690]: nit: Can you sort the array alphabetically. It looks like it was sorted before.

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (gysit) [inline mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:1354]:
```suggestion
  if (auto denormalFPMath = func.getDenormalFpMath())
```
nit: the variable names should start lowercase (below as well).

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td
Comment (gysit) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:1462]: nit: Are these flags related to the other fp_math flags above? Would it make sense to group them together?

**Review by gysit — COMMENTED** [review]
Thanks!

I added some mostly nit comments.

File: mlir/lib/Target/LLVMIR/ModuleImport.cpp
Comment (runseny) [reply mlir/lib/Target/LLVMIR/ModuleImport.cpp]: Yeah, done

File: mlir/lib/Target/LLVMIR/ModuleTranslation.cpp
Comment (runseny) [reply mlir/lib/Target/LLVMIR/ModuleTranslation.cpp]: Thanks, modified. The diff please see: [diff](https://github.com/llvm/llvm-project/compare/94bd71f97c6e35fac774611554d0069cc0e16836..c296c22b72f24c37acf37614b3f5001f1eb6475d#diff-0e9fd6478c7a83100254a3b40e2d97649a8d41c84506e1da7f8d281fc633b1f5R1354)

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td
Comment (runseny) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td]: Yeah, these flags are related to fp_math,  so i grouped these flags with other fastMath flags.

File: mlir/test/Target/LLVMIR/Import/function-attributes.ll
Comment (runseny) [reply mlir/test/Target/LLVMIR/Import/function-attributes.ll:351]: > Or is there specific reason for testing the different strings?

These different strings are the general cmd options in codegen such as [denormalFpMath related](https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/CommandFlags.cpp#L259) and [fp-contract](https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/CommandFlags.cpp#L301). So i tested all these general cmd options for functional completeness.

Comment (gysit) [reply mlir/test/Target/LLVMIR/Import/function-attributes.ll:351]: Ok! I saw there is already prior art above. 

I think the point I tried to make is that testing different values would make a lot of sense if the stored value would be an enum. Then we want to test that all enum values exist and are handled properly. In this case, we store a string so if one works all other strings should work as well (there are no separate code paths for different enum values). For maintenance reasons, I would thus keep the test set minimal if there is no extra functionality (in the sense of code coverage) that is tested.

**Review by gysit — APPROVED** [review]
LGTM

Thanks for addressing the comments! I still think there is some redundancy in terms of test cases. Feel free to trim them down if you think my argument makes sense, or otherwise keep them :).

**Comment by runseny** [comment]
> LGTM
> 
> Thanks for addressing the comments! I still think there is some redundancy in terms of test cases. Feel free to trim them down if you think my argument makes sense, or otherwise keep them :).

Yeah, you are right, thanks for this suggestion. Now i only test one string for each attribute.

**Comment by runseny** [comment]
@zero9178 Hi, i still need an extra approval, Could you help take a look?

**Review by zero9178 — APPROVED** [review]
Not super familiar with these fp functions, but given it mirrors LLVM, its their mess to figure and first and foremost :slightly_smiling_face: 
LGTM with clang-format addressed. Feel free to ping if no one merged this in a day once CI is green.

Given that some of these only have a finite set of limited values afaik, we could turn these into an enum attribute in the future. But this can be a future iteration.

> @zero9178 Hi, i still need an extra approval, Could you help take a look?

FYI, there is no such general rule in LLVM that you need to have more than one approval, although you're of course free to get more eyes on the review :slightly_smiling_face:! Sometimes the first reviewer will ask you for either more time before landing or another approval if they do not feel comfortable being the only approval, but that is not applicable here.

**Comment by runseny** [comment]
> Not super familiar with these fp functions, but given it mirrors LLVM, its their mess to figure and first and foremost 🙂 LGTM with clang-format addressed. Feel free to ping if no one merged this in a day once CI is green.
> 
> Given that some of these only have a finite set of limited values afaik, we could turn these into an enum attribute in the future. But this can be a future iteration.
> 
> > @zero9178 Hi, i still need an extra approval, Could you help take a look?
> 
> FYI, there is no such general rule in LLVM that you need to have more than one approval, although you're of course free to get more eyes on the review 🙂! Sometimes the first reviewer will ask you for either more time before landing or another approval if they do not feel comfortable being the only approval, but that is not applicable here.

Thanks for your detail explanation. Now, the CI of this mr passed, could you help merge it? thanks. @zero9178

**Pull request merged**
